File: src/AdventOfCode/Puzzles/Y2018/Day01.cs
// Copyright (c) Martin Costello, 2015. All rights reserved.
// Licensed under the Apache 2.0 license. See the LICENSE file in the project root for full license information.

namespace MartinCostello.AdventOfCode.Puzzles.Y2018;

/// <summary>
/// A class representing the puzzle for <c>https://adventofcode.com/2018/day/1</c>. This class cannot be inherited.
/// </summary>
[Puzzle(2018, 01, "Chronal Calibration", RequiresData = true, IsSlow = true)]
public sealed class Day01 : Puzzle<int, int>
{
    /// <summary>
    /// Calculates the resulting frequency after applying the specified sequence.
    /// </summary>
    /// <param name="sequence">A sequence of frequency shifts to apply.</param>
    /// <returns>
    /// The resulting frequency from applying the sequence from <paramref name="sequence"/>.
    /// </returns>
    public static int CalculateFrequency(IList<int> sequence)
        => CalculateFrequencyWithRepetition(sequence).Frequency;

    /// <summary>
    /// Calculates the resulting frequency after applying the specified sequence.
    /// </summary>
    /// <param name="sequence">A sequence of frequency shifts to apply.</param>
    /// <returns>
    /// The resulting frequency from applying the sequence from <paramref name="sequence"/>.
    /// </returns>
    public static (int Frequency, int FirstRepeat) CalculateFrequencyWithRepetition(IList<int> sequence)
    {
        int current = 0;
        int? frequency = null;
        int? firstRepeat = null;

        var history = new List<int>(sequence.Count)
        {
            current,
        };

        bool isInfinite = TendsToInfinity(sequence);

        do
        {
            foreach (int shift in sequence)
            {
                int previous = current;
                current += shift;

                if (!firstRepeat.HasValue && history.Contains(current) && history[^2] != previous)
                {
                    firstRepeat = current;
                }

                history.Add(current);
            }

            if (!frequency.HasValue)
            {
                frequency = current;
            }
        }
        while (!firstRepeat.HasValue && !isInfinite);

        return (frequency.Value, firstRepeat ?? frequency.Value);
    }

    /// <inheritdoc />
    protected override async Task<PuzzleResult> SolveCoreAsync(string[] args, CancellationToken cancellationToken)
    {
        var sequence = await ReadResourceAsNumbersAsync<int>(cancellationToken);

        (Solution1, Solution2) = CalculateFrequencyWithRepetition(sequence);

        if (Verbose)
        {
            Logger.WriteLine($"The resulting frequency is {Solution1:N0}.");
            Logger.WriteLine($"The first repeated frequency is {Solution2:N0}.");
        }

        return Result();
    }

    /// <summary>
    /// Returns whether the specified sequences tends towards infinity.
    /// </summary>
    /// <param name="sequence">The sequence of integers.</param>
    /// <returns>
    /// <see langword="true"/> if the sequence tends towards positive or negative infinity.
    /// </returns>
    private static bool TendsToInfinity(IList<int> sequence)
        => Math.Abs(sequence.Sum(Math.Sign)) == sequence.Count((p) => p is not 0);
}

File: src/AdventOfCode.Site/scripts/ts/puzzles/2018/Day01.ts
// Copyright (c) Martin Costello, 2015. All rights reserved.
// Licensed under the Apache 2.0 license. See the LICENSE file in the project root for full license information.

import { Puzzle2018 } from './Puzzle2018';

export class Day01 extends Puzzle2018 {
    frequency: number;
    firstRepeatedFrequency: number;

    override get name() {
        return 'Chronal Calibration';
    }

    override get day() {
        return 1;
    }

    static calculateFrequency(sequence: number[]) {
        const [frequency] = Day01.calculateFrequencyWithRepetition(sequence);
        return frequency;
    }

    static calculateFrequencyWithRepetition(sequence: number[]): [frequency: number, firstRepeat: number] {
        let current = 0;
        let frequency: number | null = null;
        let firstRepeat: number | null = null;

        const history: number[] = [];
        history.push(current);

        const tendsToInfinity = (sequence: number[]): boolean => {
            const x = sequence.map((p) => Math.sign(p)).reduce((x, y) => x + y, 0);
            const y = sequence.filter((p) => p !== 0, 0).length;

            return Math.abs(x) === y;
        };

        const isInfinite = tendsToInfinity(sequence);

        do {
            for (const shift of sequence) {
                let previous = current;
                current += shift;

                if (firstRepeat === null && history.includes(current) && history[history.length - 2] !== previous) {
                    firstRepeat = current;
                }

                history.push(current);
            }

            if (frequency === null) {
                frequency = current;
            }
        } while (firstRepeat === null && !isInfinite);

        return [frequency, firstRepeat !== null ? firstRepeat : frequency];
    }

    override solveCore(_: string[]) {
        const sequence = this.readResourceAsNumbers();

        [this.frequency, this.firstRepeatedFrequency] = Day01.calculateFrequencyWithRepetition(sequence);

        console.info(`The resulting frequency is ${this.frequency}.`);
        console.info(`The first repeated frequency is ${this.firstRepeatedFrequency}.`);

        return this.createResult([this.frequency, this.firstRepeatedFrequency]);
    }
}
